### PR TITLE
Add parser to read TLEs from Multi Mission Administrative Messages

### PR DESCRIPTION
**Pull request by pnuu**
This PR adds a parser that can read TLE data from MMAM XML files.

 - [x] Tests added 
 - [x] Tests passed 
 - [x] Passes ``flake8 pyorbital`` 
 - [x] Fully documented

### CODE REVIEW

**Review by mraspaud — COMMENTED** [review]
Could you add the new xml reading to the TLE class?

**Comment by pnuu** [comment]
I'll be ignoring the Deepcode warning about using `next()` without an optional default value.

File: pyorbital/tlefile.py [hunk 1]
```diff
@@ -74,11 +74,14 @@ def read_platform_numbers(in_upper=False, num_as_int=False):
             parts = row.split()
             if len(parts) < 2:
                 continue
+            # The satellite name might have whitespace
+            platform = ' '.join(parts[:-1])
+            num = parts[-1]
```
Comment (mraspaud) [inline pyorbital/tlefile.py:79]: This doesn't work I think, examples from http://celestrak.com/NORAD/elements/weather.txt :
DMSP 5D-3 F16 (USA 172) 
EWS-G1 (GOES 13)
I'm not saying it was working better before though...

File: pyorbital/tlefile.py [hunk 2]
```diff
@@ -366,49 +397,68 @@ def read_tle_files(self):
...
-        for row in raw_data:
-            if row.startswith('1 '):
-                line1 = row
-            elif row.startswith('2 '):
-                line2 = row
-            else:
+
+def _parse_tles_for_downloader(item, open_func):
+    return [Tle('', tle_file=io.StringIO(tle)) for tle in
+            _get_tles_from_uris(item, open_func, platform='', only_first=False)]
```
Comment (mraspaud) [inline pyorbital/tlefile.py:417]: why the call to StringIO here? isn't open_func taking care of that?

File: pyorbital/tlefile.py [hunk 3]
```diff
@@ -366,49 +397,68 @@ def read_tle_files(self):
...
+    root = tree.getroot()
+    data = []
+    for nav in root.findall(".//navigation"):
+        data.append(nav.find(".//line-1").text)
+        data.append(nav.find(".//line-2").text)
+
+    return "\n".join(data)
+
+
+def _grouper(n, iterable, fillvalue=None):
```
Comment (mraspaud) [inline pyorbital/tlefile.py:457]: The function name could be improved, maybe matching the docstring more?

File: pyorbital/tlefile.py [hunk 4]
```diff
@@ -366,49 +397,68 @@ def read_tle_files(self):
...
-            elif row.startswith('2 '):
-                line2 = row
-            else:
+
+def _parse_tles_for_downloader(item, open_func):
+    return [Tle('', tle_file=io.StringIO(tle)) for tle in
+            _get_tles_from_uris(item, open_func, platform='', only_first=False)]
+
+
+def collect_fnames(paths):
```
Comment (mraspaud) [inline pyorbital/tlefile.py:420]:
```suggestion
def collect_filenames(paths):
```

**Review by mraspaud — COMMENTED** [review]
Thanks for adding MMAM functionality and taking the time to refactor!

I think the tests are missing the case of passing the MMAM to the TLE class instance directly, right?
Otherwise, a few comments inline

File: pyorbital/tlefile.py (hunk 1 above)
Comment (pnuu) [reply pyorbital/tlefile.py:79]: It will. The `parts[-1]` is the platform number defined in `platforms.txt`.

File: pyorbital/tlefile.py (hunk 2 above)
Comment (pnuu) [reply pyorbital/tlefile.py:417]: That was the only way of distinguishing between a filename `str` and data already read from the XML file as `str` in `_get_uris_and_open_func()`. Suggestions for a better way are wellcome.

File: pyorbital/tlefile.py (hunk 3 above)
Comment (pnuu) [reply pyorbital/tlefile.py:457]: Will do. The naming was taken from `itertools` examples.

Comment (pnuu) [reply pyorbital/tlefile.py]: Done.

**Comment by pnuu** [comment]
> I think the tests are missing the case of passing the MMAM to the TLE class instance directly, right?

Like this one? :grin: 
https://github.com/pytroll/pyorbital/blob/89d693516431011fbc5e6b927b145f5b0629610c/pyorbital/tests/test_tlefile.py#L120-L130

**Comment by mraspaud** [comment]
Ah, sorry, missed it!

**Pull request merged**